### PR TITLE
profile refactor

### PR DESCRIPTION
**Pull request by bqpd**
@whoburg, ready for review.

## Results
With `gas_hale`, latest commit / master

| Operation | Time (%) |  Biggest cause of slowness in refactor |
|---|---|---|
| Parse student's code  | 50/47  | ???  | 
| Rename model  | 25/0  | slow NomialData generation | 
| Generate GP | 22/14 | slow Nomial generation | 
| Solve | 10 | the solver (mosek here: cvxopt is 10x slower on my computer) | 
| Compile result | 20/0 | calculating var_senss for each PosyIneq separately | 

## Conclusion
We added two new parsing steps: one a proper varkey renaming, the other a recursive result compilation. These are slow: this model changes from spending 86% of its time in GPkit to spending 92% of its time there (or 37/53% with cvxopt). Honestly, the big problem is how slow we are on the user-given code.

I'm still in favor of merging the refactor in over master.

### CODE REVIEW

**Comment by bqpd** [comment]
Update after latest commit: 

| Operation | Time (%) |  Biggest cause  | Potential fix |
|---|---|---|---|
| Parse student's code  | 44/50  | slow nomial generation  |  |
| Rename model  | 36/25  | slow nomial generation |  |
| Generate GP | 44/28 | slow nomial generation |  |
| Solve | 100 | down to the solver |  |
| Compile result | 20 | ConstraintSet.sens_from_dual | 'sens_from_dual` needs a profiling |

**Comment by bqpd** [comment]
Update after latest commit: 

| Operation | Time (%) |  Biggest cause  |
|---|---|---|---|
| Parse student's code  | 50  | slow nomial generation  | 
| Rename model  | 25  | slow nomial and varkey generation | 
| Generate GP | 28 | slow nomial generation | 
| Solve | 100 | down to the solver | 
| Compile result | 20 | ConstraintSet.sens_from_dual |

**Comment by whoburg** [comment]
It turns out that `Variable.__init__` is surprisingly slow. It takes about a full second to create 4000 variables. The script below takes 22-23 seconds to run on my computer.

I'd really like to see the refactor improve on master timing wise -- maybe Variable creation is a place we can cut out some time?

```python
﻿﻿import cProfile, pstats, StringIO
pr = cProfile.Profile()
pr.enable()

from gpkit import Variable
for i in range(100000):
    x = Variable(r"\theta_{something}", 55, "knots", "an interesting speed")

pr.disable()
s = StringIO.StringIO()
sortby = 'tottime'
ps = pstats.Stats(pr, stream=s).sort_stats(sortby)
ps.print_stats()
with open("pout.txt", "w") as outfile:
    outfile.write(s.getvalue())
```

**Comment by bqpd** [comment]
Those are going to be speed improvements that could just as easily be applied to master, so I'd propose merging this (and the refactor, pending interactives tests) and then making a goal for 0.4.1 to be speed improvements.

**Comment by bqpd** [comment]
The Variables init cost is 50% in VarKey creation, 20% NomialData init, 14% pints in Signomial.__init__.

In VarKey creation, 80% pints.

In NomialData init, 17% pints, 19% check for nonpostive cs, 49% simplify.

In simplify, 45% pints.

**Comment by whoburg** [comment]
`make_unitstr` takes awhile too

**Comment by bqpd** [comment]
That's pints.

**Comment by bqpd** [comment]
Whatever they do for string formatting takes forever.

**Comment by whoburg** [comment]
yeah goodness, are they regenerating the unit registry every time?

```python
    def __format__(self, spec):
        spec = spec or self.default_format

        if '~' in spec:
            units = UnitsContainer(dict((self._REGISTRY._get_symbol(key), value)
                                   for key, value in self.units.items()))
            spec = spec.replace('~', '')
        else:
            units = self.units

        return '%s %s' % (format(self.magnitude, remove_custom_flags(spec)),
                          format(units, spec))
```

File: gpkit/nomials/nomial_math.py [hunk 1]
```diff
@@ -118,9 +118,9 @@ def __init__(self, exps=None, cs=1, require_positive=True, simplify=True,
             self.exp = self.exps[0]
             self.c = self.cs[0]
 
-    def to(self, arg):
+    def convert_to(self, arg):
         "Convert to in the pint units sense"
-        return Signomial(self.exps, self.cs.to(arg).tolist())
+        self.cs = self.cs.to(arg)
```
Comment (whoburg) [inline gpkit/nomials/nomial_math.py]: shall we leave the `to` method for users, and add `convert_to` as the in place version as well?

Comment (whoburg) [inline gpkit/nomials/nomial_math.py]: (`to` is a nice name because it follows pint convention)

Comment (bqpd) [inline gpkit/nomials/nomial_math.py]: Sounds good!

**Comment by bqpd** [comment]
It looks like the "~" check is looking for the shortened unit names.

File: gpkit/tests/t_constraints.py [hunk 2]
```diff
@@ -30,8 +30,8 @@ def test_additive_scalar(self):
         c2 = 1 >= 5*x + 0.5
         self.assertEqual(type(c1), PosynomialInequality)
         self.assertEqual(type(c2), PosynomialInequality)
-        self.assertEqual(c1.posylt1_rep.cs, c2.posylt1_rep.cs)
-        self.assertEqual(c1.posylt1_rep.exps, c2.posylt1_rep.exps)
+        self.assertEqual(c1.as_posyslt1()[0].cs, c2.as_posyslt1()[0].cs)
+        self.assertEqual(c1.as_posyslt1()[0].exps, c2.as_posyslt1()[0].exps)
```
Comment (whoburg) [inline gpkit/tests/t_constraints.py]: shall we use your 
`var, = thing_to_unpack`
trick here to also assert length 1?

Comment (bqpd) [inline gpkit/tests/t_constraints.py]: Sure!

**Comment by whoburg** [comment]
I'm ok with merging this to the refactor.

**Comment by bqpd** [comment]
Just responded to your comments, will merge once tests pass. :)

**Comment by whoburg** [comment]
sounds great.

**Pull request merged**